flowbius/src/main/kotlin/com/trello/flowbius/FlowConnectables.kt

```diff
@@ -19,21 +21,38 @@ private class FlowConnectable<I, O>(
   private val mapper: FlowTransformer<I, O>
 ) : Connectable<I, O> {
   override fun connect(output: Consumer<O>): Connection<I> {
-    val sharedFlow = MutableSharedFlow<I>(replay = 1, extraBufferCapacity = Int.MAX_VALUE)
+    val sharedFlow = MutableSharedFlow<I>(replay = 0, extraBufferCapacity = Int.MAX_VALUE)
 
-    val job = GlobalScope.launch(Dispatchers.Unconfined + context, start = CoroutineStart.ATOMIC) {
+    val connectableJob = Job()
+    val connectableScope = CoroutineScope(connectableJob + Dispatchers.Unconfined + context)
+
+    connectableScope.launch(start = CoroutineStart.ATOMIC) {
       sharedFlow
-        .run { mapper(this) }
+        .run {
+          mapper(this)
+        }
         .collect { output.accept(it) }
     }
 
     return object : Connection<I> {
+      private val exceptionHandler = CoroutineExceptionHandler { _, throwable ->
+        if (throwable is NoSuchElementException) {
+          // flow was cancelled before receiving a subscriber, no need to handle
+        } else {
+          throw throwable
+        }
+      }
+
       override fun accept(value: I) {
-        sharedFlow.tryEmit(value)
+        connectableScope.launch(exceptionHandler) {
+          // Wait for a subscription before emitting!
+          sharedFlow.subscriptionCount.first { it > 0 }
```

Comment:
Do we ever get into a state where a listener unsubscribes but plans to re-subscribe? I don't think we do in practice but IDK if it is prevented altogether by the code. If we can, I wonder what happens in that case, I guess the events would be buffered until a reconnection?

Reply:
I don't think that would be an issue here. The only subscriber to `sharedFlow` is the `collect` within `FlowConnectables` and we manage that subscription internally in `FlowConnectables`. The only externally open path for cancelling the subscription would be cancelling `connectableJob` via `dispose` which would also cancel the coroutine within `accept`, but this should be preferred as it's part of the loop stopping.

Look at it though, there is one exception I'm seeing to what I just mentioned. If `mapper` throws an exception, the whole job is cancelled, no subsequent emissions will be collected. This is existing behavior though. I'm not 100% sure if this was an intentional design decision or a separate issue. It looks to be caught/handled in `FlowMobius.addTransformer`, but could still be problematic if `asConnectable` is used directly, which is public via the API.